hydroflow/src/compiled/pull/half_join_state/fold_from.rs

```diff
@@ -0,0 +1,48 @@
+use std::collections::hash_map::Entry::*;
+
+use rustc_hash::FxHashMap;
+
+use crate::util::clear::Clear;
+
+pub struct HalfJoinStateFoldFrom<K, A> {
+    pub table: FxHashMap<K, A>,
+}
+
+impl<K, A> Default for HalfJoinStateFoldFrom<K, A> {
+    fn default() -> Self {
+        Self {
+            table: Default::default(),
+        }
+    }
+}
+
+impl<K, A> Clear for HalfJoinStateFoldFrom<K, A> {
+    fn clear(&mut self) {
+        self.table.clear()
+    }
+}
+
+impl<K, A> HalfJoinStateFoldFrom<K, A>
+where
+    K: Eq + std::hash::Hash,
+{
+    pub fn fold_into<V, X>(
+        &mut self,
+        iter: impl Iterator<Item = (K, V)>,
+        mut fold: impl FnMut(&mut A, V) -> X,
+        mut from: impl FnMut(V) -> A,
+    ) {
+        for (k, v) in iter {
+            let entry = self.table.entry(k);
+
+            match entry {
+                Occupied(mut e) => {
+                    fold(e.get_mut(), v);
+                }
+                Vacant(e) => {
+                    e.insert(from(v));
```

Comment:
This is the only diff from `fold.rs` correct? Is there a DRYer way to do this in Rust?

Similar comments on the other cases here -- a lot of repeated boilerplate.

Reply:
I agree that fold/reduce/mutliset2/fold_from are all very similar. There are ways to factor it out but I think it would make the code less clear.

Additionally, there will be a follow up PR to this one that will try to unify join/join_multiset into the fold_fused approach and for that one I think this will likely require some re-jigging so I would like to defer it until then.